Add refined version for the to_ndarray_vfl() function #77

noah-kreutzer-frequenz opened the pull request:
According to issue #77 a refined version of to_ndarray_vfl() to_ndarray_vfl_refined() was added to the repo. to_ndarray_vfl_refined() in comparision to the base function takes the validity_times to filter for in form of a list datetime Time objects instead of a list of timedelta relative to the request.

noah-kreutzer-frequenz committed:
Add refined version for the to_ndarray_vfl() function #77

According to issue #77 a refined version of to_ndarray_vfl() to_ndarray_vfl_refined() was added to the repo.
to_ndarray_vfl_refined() in comparision to the base function takes the validity_times to filter for
in form of a list of datetime objects instead of a list of timedelta relative to the request.

Signed-off-by: noah-kreutzer-frequenz <noah.kreutzer@frequenz.com>

cwasicki commented on py/frequenz/client/weather/_types.py:
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
             # and features
-            if array.shape[0] != len(validity_times_indexes):
+            if validity_times is not None and array.shape[0] != len(validity_times):
                 print(
```
`print` should be replaced by the logger (can be separate PR).

cwasicki commented on py/frequenz/client/weather/_types.py:
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
             # Check if the array shape matches the number of filtered times, locations
             # and features
-            if array.shape[0] != len(validity_times_indexes):
+            if validity_times is not None and array.shape[0] != len(validity_times):
```
This change is not related to adding the refined function, right? Then it should be a separate commit at least.

cwasicki commented on py/frequenz/client/weather/_types.py:
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
-                    f"Warning:  The count of validity times in the "
-                    f"array({array.shape[0]}) does not match the expected time "
-                    f"filter count ({validity_times_indexes}."
+                    (
```
why these additional `(`?

cwasicki commented on py/frequenz/client/weather/_types.py:
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
                 )
-            if array.shape[2] != len(feature_indexes):
+            if features is not None and array.shape[2] != len(features):
+                print(
```
Why are all of these checks only warnings, is there a reasonable situation where this can fail?

noah-kreutzer-frequenz replied:
@cwasicki I thought its a good check if you actually requested the stream for certain configuration a while ago. Before wondering why some features, validity_times or locations are not shown in the output. This warning should inform you that the features are not in part of the stream.
Maybe a clearer better Warning should be implemented

cwasicki commented on py/frequenz/client/weather/_types.py:
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
+        return array
+
+    # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+    def to_ndarray_vlf_refined(
```
It's not clear what refined means. Why do we even need both? We can just change the validity_times type in the original `to_ndarray_vlf`  as suggested in #77

cwasicki commented on py/frequenz/client/weather/_types.py:
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
+    # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+    def to_ndarray_vlf_refined(
+        self,
+        validity_times: list[dt.time] | None = None,
```
This should be datetime as time beyond 24h is ambiguous

noah-kreutzer-frequenz replied:
@cwasicki can we have a quick call? I'm not sure if I got what kind of change you wanted.

cwasicki commented on py/frequenz/client/weather/_types.py:
```diff
@@ -263,19 +263,161 @@ def to_ndarray_vlf(
+        if not self._forecasts_pb.location_forecasts:
+            raise ValueError("Forecast data is missing or invalid.")
+
+        try:
```
As mentioned above I suggest to have only one function, so just a general suggestion: There is a lot of duplicated code here, it would be better to refactor it to share as much common code as possible.

cwasicki commented on py/frequenz/client/weather/_types.py:
```diff
@@ -263,19 +256,21 @@ def to_ndarray_vlf(
-                    f"Warning:  The count of validity times in the "
-                    f"array({array.shape[0]}) does not match the expected time "
-                    f"filter count ({validity_times_indexes}."
+                    (
```
Do we need these extra layers of parantheses? Also, `print`s should be replaced with logger.

noah-kreutzer-frequenz replied:
I would rework the print statements in a different PR, i already wrote an issue for that

noah-kreutzer-frequenz committed:
Restructure Tests

Restructure the test_types.py to reuse certain test variables and aligne with the pylint requirements

Signed-off-by: noah-kreutzer-frequenz <noah.kreutzer@frequenz.com>
